File: src/main/java/com/springdebugger/enricher/PropertyPrecedenceEnricher.java
package com.springdebugger.enricher;

import com.springdebugger.model.DiagnosisCard;
import com.springdebugger.model.Phase;
import com.springdebugger.model.RawSignal;

import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Property-precedence enrichment (M9, completing the Actuator layer). When a runtime
 * configuration error references a property key and the app is still alive with Actuator
 * exposed, this reports which property source actually supplies the effective value via
 * /actuator/env/{key}. This is the consumer for {@link ActuatorReader#effectivePropertySource}.
 *
 * <p>Strictly additive: it appends a clause when it can confirm the source and otherwise
 * returns the card unchanged. Confidence is left as-is (this adds context, not certainty).
 */
public final class PropertyPrecedenceEnricher implements Enricher {

    /** A dotted, lower-case Spring property key such as spring.datasource.url or app.timeout. */
    private static final Pattern PROPERTY_KEY = Pattern.compile("[a-z][a-z0-9]*(?:\\.[a-z0-9-]+)+");

    @Override
    public DiagnosisCard enrich(DiagnosisCard card, RawSignal signal, EnrichmentContext context) {
        if (card == null || context == null) return card;
        if (signal.getPhase() != Phase.RUNTIME) return card;

        String key = firstPropertyKey(signal.getDeepestCausedByMessage());
        if (key == null) key = firstPropertyKey(card.getDiagnosisSentence());
        if (key == null) return card;

        Optional<String> env = safeGet(context, "/actuator/env/" + key);
        if (env.isEmpty()) return card;

        Optional<String> source = ActuatorReader.effectivePropertySource(env.get());
        if (source.isEmpty()) return card;

        return new DiagnosisCard(
                card.getRuleId(),
                card.getPhase(),
                card.getDiagnosisSentence()
                        + " The effective value of '" + key + "' comes from " + source.get() + ".",
                card.getFixSentence(),
                card.getConfidence(),
                card.getExcerpt());
    }

    static String firstPropertyKey(String text) {
        if (text == null) return null;
        Matcher m = PROPERTY_KEY.matcher(text);
        return m.find() ? m.group() : null;
    }

    private Optional<String> safeGet(EnrichmentContext context, String path) {
        try {
            return context.httpGet(path);
        } catch (RuntimeException e) {
            return Optional.empty();
        }
    }
}

File: vscode-extension/src/engine/property-precedence-enricher.ts
// Port of com.springdebugger.enricher.PropertyPrecedenceEnricher. When a runtime config error names
// a property key and the app is alive with Actuator, report which property source actually supplies
// the effective value via /actuator/env/{key}. Strictly additive; confidence left unchanged.
import { Enricher, EnrichmentContext } from './enricher';
import { DiagnosisCard, RawSignal } from './models';
import { effectivePropertySource } from './actuator-reader';

const PROPERTY_KEY = /[a-z][a-z0-9]*(?:\.[a-z0-9-]+)+/;

export class PropertyPrecedenceEnricher implements Enricher {
  async enrich(card: DiagnosisCard, signal: RawSignal, context: EnrichmentContext): Promise<DiagnosisCard> {
    if (signal.phase !== 'RUNTIME') return card;

    let key = firstPropertyKey(signal.deepestCausedByMessage);
    if (key === null) key = firstPropertyKey(card.diagnosisSentence);
    if (key === null) return card;

    const env = await safeGet(context, '/actuator/env/' + key);
    if (env === null) return card;

    const source = effectivePropertySource(env);
    if (source === null) return card;

    return {
      ruleId: card.ruleId,
      phase: card.phase,
      diagnosisSentence: `${card.diagnosisSentence} The effective value of '${key}' comes from ${source}.`,
      fixSentence: card.fixSentence,
      confidence: card.confidence,
      excerpt: card.excerpt,
    };
  }
}

export function firstPropertyKey(text: string | null): string | null {
  if (text === null) return null;
  const m = PROPERTY_KEY.exec(text);
  return m ? m[0] : null;
}

async function safeGet(context: EnrichmentContext, path: string): Promise<string | null> {
  try {
    return await context.httpGet(path);
  } catch {
    return null;
  }
}
